[14.0] stock_available_to_promise_release & stock_release_channel: refactor releasing

jbaudoux opened the pull request:
## stock_available_to_promise: refactor releasing

Regroup concepts in methods called on the recordset. One method to split the moves. One method for making the backorder.
This will prevent stock_release_channel to spawn a job for each move. A single job is now spawn to compute the channels. 

## stock_release_channel: do not copy release channel

As a picking return is created by copying the delivery, do not copy the release channel as we don't want it on a return picking (reception picking). 

Ensure the release channel assignment job is created after the move is assigned to the new picking. The job is now created at the end of `_assign_picking`.

cc @sebalix @lmignon @mt-software-de

jbaudoux committed:
s_available_to_promise_release: refactor release

Regroup concepts and call methods on recordset.
This improves code readability and will allow stock_release_channel to
spawn only one job instead of one job for each move.

jbaudoux committed:
stock_available_to_promise_release: adding BCIM as author

Adding BICM as co-designer of this module, and author of the initial
module in v10.0 from which this has been partly inspired.

mt-software-de commented on stock_available_to_promise_release/models/stock_move.py:
```diff
@@ -422,12 +423,16 @@ def _release_split(self, remaining_qty):
+            for new_move_val in new_move_vals:
+                new_move_val["state"] = "draft"
+                if "picking_id" in new_move_val:
+                    del new_move_val["picking_id"]
```
Maybe we move than also Line 421 and 422 into the if?

jbaudoux committed:
stock_release_channel: do not copy release channel

As a picking return is created by copying the delivery, do not copy the
release channel as we don't want it on a return picking.
Ensure the release channel assignment job is created after the move is
assigned to the new picking.

jbaudoux commented:
Todo: add test for return picking

mt-software-de reviewed (approved):
LGTM

lmignon reviewed (approved):
Great change. The code is easier to read and the functional scope of methods is more clearly defined. LGTM (Code review)

jbaudoux commented:
rebased after merge of conflicting PR https://github.com/OCA/wms/pull/556

jbaudoux commented:
@TDu Can you review this one?

jbaudoux commented:
The missing test has been added

mt-software-de reviewed (approved):
LGTM

simahawk commented:
Sounds good. @sebalix  pls have a look :)

jbaudoux commented:
/ocabot merge minor
